components/renku_data_services/crc/db.py

```diff
@@ -487,35 +497,67 @@ async def get_resource_pool_users(
+                stmt = (
+                    select(schemas.ResourcePoolORM)
+                    .where(schemas.ResourcePoolORM.id == resource_pool_id)
+                    .options(selectinload(schemas.ResourcePoolORM.users))
+                )
+                if keycloak_id is not None:
+                    stmt = stmt.join(schemas.ResourcePoolORM.users, isouter=True).where(
+                        or_(
+                            schemas.UserORM.keycloak_id == keycloak_id,
+                            schemas.ResourcePoolORM.public == true(),
+                            schemas.ResourceClassORM.default == true(),
                         )
-                    return [user.dump() for user in rp.users]
-                else:
-                    stmt_usr = select(schemas.UserORM)
-                    if keycloak_id is not None:
-                        stmt_usr = stmt_usr.where(schemas.UserORM.keycloak_id == keycloak_id)
-                    res_usr = await session.execute(stmt_usr)
-                    orms = res_usr.scalars().all()
-                    return [orm.dump() for orm in orms]
+                    )
+                res = await session.execute(stmt)
+                rp = res.scalars().first()
+                if rp is None:
+                    raise errors.MissingResourceError(
+                        message=f"Resource pool with id {resource_pool_id} does not exist"
+                    )
+                specific_user: base_models.User | None = None
+                if keycloak_id:
+                    specific_user_res = (
+                        await session.execute(select(schemas.UserORM).where(schemas.UserORM.keycloak_id == keycloak_id))
+                    ).scalar_one_or_none()
+                    specific_user = None if not specific_user_res else specific_user_res.dump()
+                allowed: List[base_models.User] = []
+                disallowed: List[base_models.User] = []
+                if rp.default:
+                    disallowed_stmt = select(schemas.UserORM).where(schemas.UserORM.no_default_access == true())
+                    if keycloak_id:
+                        disallowed_stmt = disallowed_stmt.where(schemas.UserORM.keycloak_id == keycloak_id)
+                    disallowed_res = await session.execute(disallowed_stmt)
+                    disallowed = [user.dump() for user in disallowed_res.scalars().all()]
+                elif rp.public and not rp.default:
+                    if specific_user:
+                        allowed = [specific_user]
+                elif not rp.public and not rp.default:
+                    allowed = [user.dump() for user in rp.users]
+                return RespositoryUsers(rp.id, allowed, disallowed)
+
+    @_only_admins
+    async def get_users(
+        self,
+        *,
+        api_user: base_models.APIUser,
+        id: str | None = None,
+    ) -> List[base_models.User]:
+        """Get users from the users-resource pools access lists."""
```

Comment:
this seems to just get all users, nothing with access lists.

Reply:
So the `UserORM` object is literally just a list of users that is joined with another table `resource_pools_users` to get ACLs. With the addition of a dedicated table of users synced with Keycloak this table is only storing the users that have (or are prevented from having) access to a resource pool. I will update the `UserORM` object to have a name that reflects this.